{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upload sequencing data to SRA\n",
    "This Python Jupyter notebook uploads the sequencing data to the NIH [Sequence Read Archive](https://www.ncbi.nlm.nih.gov/sra), or SRA."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create BioProject and BioSamples\n",
    "The first step was done manually to create the BioProject and BioSamples.\n",
    "Note that for new future uploads you may be able to use the existing BioProject.\n",
    "\n",
    "To create these, I went to the [Sequence Read Archive](https://www.ncbi.nlm.nih.gov/sra) and signed in using the box at the upper right of the webpage, and then went to the [SRA Submission Portal](https://submit.ncbi.nlm.nih.gov/subs/sra/).\n",
    "I then manually completed the first five steps, which define the project and samples."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create submission sheet\n",
    "The sixth step is to create the submission sheet in `*.tsv` format, which is done by the following code.\n",
    "\n",
    "First, import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the configuration for the analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../config.yaml') as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the PacBio runs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading PacBio runs from ../data/PacBio_runs.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>run</th>\n",
       "      <th>subreads</th>\n",
       "      <th>ccs_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lib1</td>\n",
       "      <td>200415_A</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...</td>\n",
       "      <td>../results/ccs/lib1_200415_A_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lib1</td>\n",
       "      <td>200415_B</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...</td>\n",
       "      <td>../results/ccs/lib1_200415_B_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lib2</td>\n",
       "      <td>200415_A</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...</td>\n",
       "      <td>../results/ccs/lib2_200415_A_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lib2</td>\n",
       "      <td>200415_B</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...</td>\n",
       "      <td>../results/ccs/lib2_200415_B_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  library       run                                           subreads  \\\n",
       "0    lib1  200415_A  /fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...   \n",
       "1    lib1  200415_B  /fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...   \n",
       "2    lib2  200415_A  /fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...   \n",
       "3    lib2  200415_B  /fh/fast/bloom_j/SR/ngs/pacbio/200415_TylerSta...   \n",
       "\n",
       "                                    ccs_file  \n",
       "0  ../results/ccs/lib1_200415_A_ccs.fastq.gz  \n",
       "1  ../results/ccs/lib1_200415_B_ccs.fastq.gz  \n",
       "2  ../results/ccs/lib2_200415_A_ccs.fastq.gz  \n",
       "3  ../results/ccs/lib2_200415_B_ccs.fastq.gz  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pacbio_runs_file = os.path.join('../', config['pacbio_runs'])\n",
    "\n",
    "print(f\"Reading PacBio runs from {pacbio_runs_file}\")\n",
    "\n",
    "pacbio_runs = (\n",
    "    pd.read_csv(pacbio_runs_file)\n",
    "    .assign(ccs_file=lambda x: f\"../{config['ccs_dir']}/\" + x['library'] + '_' + x['run'] + '_ccs.fastq.gz')\n",
    "    )\n",
    "\n",
    "pacbio_runs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next make submission entries for the PacBio CCSs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_id</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename_fullpath</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib1_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib1_200415_A_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib1_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib1_200415_B_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib2_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib2_200415_A_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib2_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib2_200415_B_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_name        library_id  \\\n",
       "0  PacBio_CCSs  lib1_PacBio_CCSs   \n",
       "1  PacBio_CCSs  lib1_PacBio_CCSs   \n",
       "2  PacBio_CCSs  lib2_PacBio_CCSs   \n",
       "3  PacBio_CCSs  lib2_PacBio_CCSs   \n",
       "\n",
       "                                               title     library_strategy  \\\n",
       "0  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "1  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "2  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "3  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "\n",
       "  library_source   library_selection library_layout     platform  \\\n",
       "0      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "1      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "2      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "3      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "\n",
       "  instrument_model                                 design_description  \\\n",
       "0    PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "1    PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "2    PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "3    PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "\n",
       "  filetype                          filename_fullpath  \n",
       "0    fastq  ../results/ccs/lib1_200415_A_ccs.fastq.gz  \n",
       "1    fastq  ../results/ccs/lib1_200415_B_ccs.fastq.gz  \n",
       "2    fastq  ../results/ccs/lib2_200415_A_ccs.fastq.gz  \n",
       "3    fastq  ../results/ccs/lib2_200415_B_ccs.fastq.gz  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pacbio_submissions = (\n",
    "    pacbio_runs\n",
    "    .assign(\n",
    "        sample_name='PacBio_CCSs',  # BioSample created in SRA wizard\n",
    "        library_id=lambda x: x['library'] + '_PacBio_CCSs',  # unique library ID\n",
    "        title='PacBio CCSs linking variants to barcodes for SARS-CoV-2 RBD deep mutational scanning',\n",
    "        library_strategy='Synthetic-Long-Read',\n",
    "        library_source='SYNTHETIC',\n",
    "        library_selection='Restriction Digest',\n",
    "        library_layout='single',\n",
    "        platform='PACBIO_SMRT',\n",
    "        instrument_model='PacBio Sequel',\n",
    "        design_description='Restriction digest of plasmids carrying barcoded RBD variants',\n",
    "        filetype='fastq',\n",
    "        filename_fullpath=lambda x: x['ccs_file'],      \n",
    "        )\n",
    "    .drop(columns=pacbio_runs.columns)\n",
    "    )\n",
    "\n",
    "pacbio_submissions.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the Illumina runs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading Illumina runs from ../data/barcode_runs.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>sort_bin</th>\n",
       "      <th>concentration</th>\n",
       "      <th>date</th>\n",
       "      <th>number_cells</th>\n",
       "      <th>R1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lib1</td>\n",
       "      <td>SortSeq_bin1</td>\n",
       "      <td>SortSeq</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200416</td>\n",
       "      <td>6600000</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_0952...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lib1</td>\n",
       "      <td>SortSeq_bin2</td>\n",
       "      <td>SortSeq</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200416</td>\n",
       "      <td>3060000</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_0952...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lib1</td>\n",
       "      <td>SortSeq_bin3</td>\n",
       "      <td>SortSeq</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200416</td>\n",
       "      <td>2511000</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_0952...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lib1</td>\n",
       "      <td>SortSeq_bin4</td>\n",
       "      <td>SortSeq</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200416</td>\n",
       "      <td>2992000</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_0952...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lib2</td>\n",
       "      <td>SortSeq_bin1</td>\n",
       "      <td>SortSeq</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>200416</td>\n",
       "      <td>6420000</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_0953...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  library        sample sample_type  sort_bin  concentration    date  \\\n",
       "0    lib1  SortSeq_bin1     SortSeq         1            NaN  200416   \n",
       "1    lib1  SortSeq_bin2     SortSeq         2            NaN  200416   \n",
       "2    lib1  SortSeq_bin3     SortSeq         3            NaN  200416   \n",
       "3    lib1  SortSeq_bin4     SortSeq         4            NaN  200416   \n",
       "4    lib2  SortSeq_bin1     SortSeq         1            NaN  200416   \n",
       "\n",
       "   number_cells                                                 R1  \n",
       "0       6600000  /shared/ngs/illumina/tstarr/200427_D00300_0952...  \n",
       "1       3060000  /shared/ngs/illumina/tstarr/200427_D00300_0952...  \n",
       "2       2511000  /shared/ngs/illumina/tstarr/200427_D00300_0952...  \n",
       "3       2992000  /shared/ngs/illumina/tstarr/200427_D00300_0952...  \n",
       "4       6420000  /shared/ngs/illumina/tstarr/200427_D00300_0953...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "illumina_runs_file = os.path.join('../', config['barcode_runs'])\n",
    "\n",
    "print(f\"Reading Illumina runs from {illumina_runs_file}\")\n",
    "\n",
    "illumina_runs = pd.read_csv(illumina_runs_file)\n",
    "\n",
    "illumina_runs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next make submission entries for Illumina data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_id</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename_fullpath</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_0952...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample_name         library_id  \\\n",
       "0  expression_barcodes  lib1_SortSeq_bin1   \n",
       "1  expression_barcodes  lib1_SortSeq_bin1   \n",
       "2  expression_barcodes  lib1_SortSeq_bin1   \n",
       "3  expression_barcodes  lib1_SortSeq_bin1   \n",
       "4  expression_barcodes  lib1_SortSeq_bin1   \n",
       "\n",
       "                                               title library_strategy  \\\n",
       "0  SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "1  SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "2  SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "3  SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "4  SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "\n",
       "  library_source library_selection library_layout  platform  \\\n",
       "0      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "1      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "2      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "3      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "4      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "\n",
       "      instrument_model                 design_description filetype  \\\n",
       "0  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "1  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "2  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "3  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "4  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "\n",
       "                                   filename_fullpath  \n",
       "0  /shared/ngs/illumina/tstarr/200427_D00300_0952...  \n",
       "1  /fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...  \n",
       "2  /fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...  \n",
       "3  /fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...  \n",
       "4  /fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427...  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "illumina_submissions = (\n",
    "    illumina_runs\n",
    "    .assign(\n",
    "        sample_name=lambda x: x['sample_type'].map({'SortSeq': 'expression_barcodes',\n",
    "                                                    'TiteSeq': 'hACE2_binding_barcodes'}),\n",
    "        library_id=lambda x: x['library'] + '_' + x['sample'],\n",
    "        title=lambda x: 'SARS-CoV-2 RBD deep mutational scanning Illumina barcode sequencing for ' + x['sample'],\n",
    "        library_strategy='AMPLICON',\n",
    "        library_source='SYNTHETIC',\n",
    "        library_selection='PCR',\n",
    "        library_layout='single',\n",
    "        platform='ILLUMINA',\n",
    "        instrument_model='Illumina HiSeq 2500',\n",
    "        design_description='PCR of barcodes from RBD variants',\n",
    "        filetype='fastq',\n",
    "        filename_fullpath=lambda x: x['R1'].str.split(';'),       \n",
    "        )\n",
    "    .explode('filename_fullpath')\n",
    "    .assign(filename_fullpath=lambda x: x['filename_fullpath'].str.strip())\n",
    "    .drop(columns=illumina_runs.columns)\n",
    "    .reset_index(drop=True)\n",
    "    )\n",
    "\n",
    "illumina_submissions.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now concatenate the PacBio and Illumina submissions into tidy format (one line per file), make sure all the files exist, and also make short name versions of them that lack the path:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_id</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename_fullpath</th>\n",
       "      <th>file_exists</th>\n",
       "      <th>filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib1_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib1_200415_A_ccs.fastq.gz</td>\n",
       "      <td>True</td>\n",
       "      <td>lib1_200415_A_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib1_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib1_200415_B_ccs.fastq.gz</td>\n",
       "      <td>True</td>\n",
       "      <td>lib1_200415_B_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib2_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib2_200415_A_ccs.fastq.gz</td>\n",
       "      <td>True</td>\n",
       "      <td>lib2_200415_A_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PacBio_CCSs</td>\n",
       "      <td>lib2_PacBio_CCSs</td>\n",
       "      <td>PacBio CCSs linking variants to barcodes for S...</td>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>Restriction Digest</td>\n",
       "      <td>single</td>\n",
       "      <td>PACBIO_SMRT</td>\n",
       "      <td>PacBio Sequel</td>\n",
       "      <td>Restriction digest of plasmids carrying barcod...</td>\n",
       "      <td>fastq</td>\n",
       "      <td>../results/ccs/lib2_200415_B_ccs.fastq.gz</td>\n",
       "      <td>True</td>\n",
       "      <td>lib2_200415_B_ccs.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_0952...</td>\n",
       "      <td>True</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L001_R1_001.fast...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample_name         library_id  \\\n",
       "0          PacBio_CCSs   lib1_PacBio_CCSs   \n",
       "1          PacBio_CCSs   lib1_PacBio_CCSs   \n",
       "2          PacBio_CCSs   lib2_PacBio_CCSs   \n",
       "3          PacBio_CCSs   lib2_PacBio_CCSs   \n",
       "4  expression_barcodes  lib1_SortSeq_bin1   \n",
       "\n",
       "                                               title     library_strategy  \\\n",
       "0  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "1  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "2  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "3  PacBio CCSs linking variants to barcodes for S...  Synthetic-Long-Read   \n",
       "4  SARS-CoV-2 RBD deep mutational scanning Illumi...             AMPLICON   \n",
       "\n",
       "  library_source   library_selection library_layout     platform  \\\n",
       "0      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "1      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "2      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "3      SYNTHETIC  Restriction Digest         single  PACBIO_SMRT   \n",
       "4      SYNTHETIC                 PCR         single     ILLUMINA   \n",
       "\n",
       "      instrument_model                                 design_description  \\\n",
       "0        PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "1        PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "2        PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "3        PacBio Sequel  Restriction digest of plasmids carrying barcod...   \n",
       "4  Illumina HiSeq 2500                  PCR of barcodes from RBD variants   \n",
       "\n",
       "  filetype                                  filename_fullpath  file_exists  \\\n",
       "0    fastq          ../results/ccs/lib1_200415_A_ccs.fastq.gz         True   \n",
       "1    fastq          ../results/ccs/lib1_200415_B_ccs.fastq.gz         True   \n",
       "2    fastq          ../results/ccs/lib2_200415_A_ccs.fastq.gz         True   \n",
       "3    fastq          ../results/ccs/lib2_200415_B_ccs.fastq.gz         True   \n",
       "4    fastq  /shared/ngs/illumina/tstarr/200427_D00300_0952...         True   \n",
       "\n",
       "                                            filename  \n",
       "0                         lib1_200415_A_ccs.fastq.gz  \n",
       "1                         lib1_200415_B_ccs.fastq.gz  \n",
       "2                         lib2_200415_A_ccs.fastq.gz  \n",
       "3                         lib2_200415_B_ccs.fastq.gz  \n",
       "4  200416_lib1_FITCbin1_TGGAACAA_L001_R1_001.fast...  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submissions_tidy = (\n",
    "    pd.concat([pacbio_submissions, illumina_submissions], ignore_index=True)\n",
    "    .assign(file_exists=lambda x: x['filename_fullpath'].map(os.path.isfile),\n",
    "            filename=lambda x: x['filename_fullpath'].map(os.path.basename),\n",
    "            )\n",
    "    )\n",
    "\n",
    "assert submissions_tidy['file_exists'].all(), submissions_tidy.query('file_exists == False')\n",
    "\n",
    "submissions_tidy.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427_D00300_0952_AHFCLCBCX3/Unaligned/Project_tstarr/Sample_200416_lib1_FITCbin1/200416_lib1_FITCbin1_TGGAACAA_L001_R1_002.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "! ls /fh/fast/bloom_j/SR/ngs/illumina/tstarr/200427_D00300_0952_AHFCLCBCX3/Unaligned/Project_tstarr/Sample_200416_lib1_FITCbin1/200416_lib1_FITCbin1_TGGAACAA_L001_R1_002.fastq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_id</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>design_description</th>\n",
       "      <th>filetype</th>\n",
       "      <th>filename_fullpath</th>\n",
       "      <th>file_exists</th>\n",
       "      <th>filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/tstarr/20042...</td>\n",
       "      <td>False</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L001_R1_002.fast...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/tstarr/20042...</td>\n",
       "      <td>False</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L001_R1_003.fast...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/tstarr/20042...</td>\n",
       "      <td>False</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L001_R1_004.fast...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/tstarr/20042...</td>\n",
       "      <td>False</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L001_R1_005.fast...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>expression_barcodes</td>\n",
       "      <td>lib1_SortSeq_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/tstarr/20042...</td>\n",
       "      <td>False</td>\n",
       "      <td>200416_lib1_FITCbin1_TGGAACAA_L002_R1_001.fast...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib2_TiteSeq_16_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_095...</td>\n",
       "      <td>False</td>\n",
       "      <td>200422_s16-b1_TATCAGCA_L002_R1_002.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib2_TiteSeq_16_bin1</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_095...</td>\n",
       "      <td>False</td>\n",
       "      <td>200422_s16-b1_TATCAGCA_L002_R1_003.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib2_TiteSeq_16_bin2</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_095...</td>\n",
       "      <td>False</td>\n",
       "      <td>200422_s16-b2_TCCGTCTA_L002_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib2_TiteSeq_16_bin3</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_095...</td>\n",
       "      <td>False</td>\n",
       "      <td>200422_s16-b3_TCTTCACA_L002_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>hACE2_binding_barcodes</td>\n",
       "      <td>lib2_TiteSeq_16_bin4</td>\n",
       "      <td>SARS-CoV-2 RBD deep mutational scanning Illumi...</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SYNTHETIC</td>\n",
       "      <td>PCR</td>\n",
       "      <td>single</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PCR of barcodes from RBD variants</td>\n",
       "      <td>fastq</td>\n",
       "      <td>/shared/ngs/illumina/tstarr/200427_D00300_095...</td>\n",
       "      <td>False</td>\n",
       "      <td>200422_s16-b4_TGAAGAGA_L002_R1_001.fastq.gz</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>236 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                sample_name            library_id  \\\n",
       "5       expression_barcodes     lib1_SortSeq_bin1   \n",
       "6       expression_barcodes     lib1_SortSeq_bin1   \n",
       "7       expression_barcodes     lib1_SortSeq_bin1   \n",
       "8       expression_barcodes     lib1_SortSeq_bin1   \n",
       "9       expression_barcodes     lib1_SortSeq_bin1   \n",
       "..                      ...                   ...   \n",
       "368  hACE2_binding_barcodes  lib2_TiteSeq_16_bin1   \n",
       "369  hACE2_binding_barcodes  lib2_TiteSeq_16_bin1   \n",
       "371  hACE2_binding_barcodes  lib2_TiteSeq_16_bin2   \n",
       "373  hACE2_binding_barcodes  lib2_TiteSeq_16_bin3   \n",
       "375  hACE2_binding_barcodes  lib2_TiteSeq_16_bin4   \n",
       "\n",
       "                                                 title library_strategy  \\\n",
       "5    SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "6    SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "7    SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "8    SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "9    SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "..                                                 ...              ...   \n",
       "368  SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "369  SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "371  SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "373  SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "375  SARS-CoV-2 RBD deep mutational scanning Illumi...         AMPLICON   \n",
       "\n",
       "    library_source library_selection library_layout  platform  \\\n",
       "5        SYNTHETIC               PCR         single  ILLUMINA   \n",
       "6        SYNTHETIC               PCR         single  ILLUMINA   \n",
       "7        SYNTHETIC               PCR         single  ILLUMINA   \n",
       "8        SYNTHETIC               PCR         single  ILLUMINA   \n",
       "9        SYNTHETIC               PCR         single  ILLUMINA   \n",
       "..             ...               ...            ...       ...   \n",
       "368      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "369      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "371      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "373      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "375      SYNTHETIC               PCR         single  ILLUMINA   \n",
       "\n",
       "        instrument_model                 design_description filetype  \\\n",
       "5    Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "6    Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "7    Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "8    Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "9    Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "..                   ...                                ...      ...   \n",
       "368  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "369  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "371  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "373  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "375  Illumina HiSeq 2500  PCR of barcodes from RBD variants    fastq   \n",
       "\n",
       "                                     filename_fullpath  file_exists  \\\n",
       "5     /fh/fast/bloom_j/SR/ngs/illumina/tstarr/20042...        False   \n",
       "6     /fh/fast/bloom_j/SR/ngs/illumina/tstarr/20042...        False   \n",
       "7     /fh/fast/bloom_j/SR/ngs/illumina/tstarr/20042...        False   \n",
       "8     /fh/fast/bloom_j/SR/ngs/illumina/tstarr/20042...        False   \n",
       "9     /fh/fast/bloom_j/SR/ngs/illumina/tstarr/20042...        False   \n",
       "..                                                 ...          ...   \n",
       "368   /shared/ngs/illumina/tstarr/200427_D00300_095...        False   \n",
       "369   /shared/ngs/illumina/tstarr/200427_D00300_095...        False   \n",
       "371   /shared/ngs/illumina/tstarr/200427_D00300_095...        False   \n",
       "373   /shared/ngs/illumina/tstarr/200427_D00300_095...        False   \n",
       "375   /shared/ngs/illumina/tstarr/200427_D00300_095...        False   \n",
       "\n",
       "                                              filename  \n",
       "5    200416_lib1_FITCbin1_TGGAACAA_L001_R1_002.fast...  \n",
       "6    200416_lib1_FITCbin1_TGGAACAA_L001_R1_003.fast...  \n",
       "7    200416_lib1_FITCbin1_TGGAACAA_L001_R1_004.fast...  \n",
       "8    200416_lib1_FITCbin1_TGGAACAA_L001_R1_005.fast...  \n",
       "9    200416_lib1_FITCbin1_TGGAACAA_L002_R1_001.fast...  \n",
       "..                                                 ...  \n",
       "368        200422_s16-b1_TATCAGCA_L002_R1_002.fastq.gz  \n",
       "369        200422_s16-b1_TATCAGCA_L002_R1_003.fastq.gz  \n",
       "371        200422_s16-b2_TCCGTCTA_L002_R1_001.fastq.gz  \n",
       "373        200422_s16-b3_TCTTCACA_L002_R1_001.fastq.gz  \n",
       "375        200422_s16-b4_TGAAGAGA_L002_R1_001.fastq.gz  \n",
       "\n",
       "[236 rows x 14 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(submissions_tidy\n",
    " .query('file_exists == False')\n",
    " ['f']\n",
    " )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
